compiler/modules/print.llvm/llvm.bal

```diff
@@ -69,11 +69,68 @@ public function constInt(IntType ty, int val) returns ConstValue {
     return new ConstValue(ty, val.toString());
 }
 
+final string pINF = (1.0/0.0).toBitsInt().toHexString().toUpperAscii();
+final string nINF = (-1.0/0.0).toBitsInt().toHexString().toUpperAscii();
+final string NAN = (0.0/0.0).toBitsInt().toHexString().toUpperAscii();
+
 // Corresponds to LLVMConstReal
 public function constReal(RealType ty, float val) returns ConstValue {
-    // XXX #307 use decimal for non-special floats 
-    return new ConstValue(ty, "0x" + int:toHexString(float:toBitsInt(val)));
-
+    string valRep;
+    // Special cases
+    if val.isInfinite() {
+        if val > 0.0 {
+            valRep = "0x" + pINF;
+        }
+        else {
+            valRep = "0x" + nINF;
+        }
+    }
+    else if val.isNaN() {
+        valRep = "0x" + NAN;
+    }
+    else {
+        // General case
+        if float:abs(val).toString().length() > 7 {
+            // Resulting number has more than 7 digits so convert to hex format
+            int repInt = float:toBitsInt(val);
+            valRep = "0x" + int:toHexString(repInt).toUpperAscii();
+        }
+        else {
+            // Convert the number to scientific notation
+            float exp = float:floor(float:log10(float:abs(val))); // Value of the exponent
```

Comment:
Why can't we use the result of val.toString() here?

Reply:
It doesn't convert the value to scientific notation it just output the value as a string for example for 11.2345 it will output "11.2345" instead of "1.123450e+01" (which is the output for llvm). We can use the string output and generate the value using a combination of bunch of splits and concats but I am not sure whether it will be more easier to understand specially given we'll have lot of special cases like values less than zero (ex 0.001) and large ones.

Reply:
So LLVM is using printf `e` format and you are trying to match that, right?

The right way to solve this is to add a toExpString function to lang.float

https://github.com/ballerina-platform/ballerina-spec/issues/770#issuecomment-791258822

We don't have that yet, so we can either

1. not try to match LLVM output in this respect,
2. create a module with some library functions and implement toExpString in that

If we do 2, we should just use the native Java equivalent (which is String.format, I think)

Reply:
Fixed by option 1 because it meets the objective of being easier to read and compiles correctly (though not exactly equal to LLVM output) while remaining purely ballerina (unlike `jni.llvm`).